homeassistant/components/tesla_wall_connector/config_flow.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from tesla_wall_connector import WallConnector
+from tesla_wall_connector.exceptions import WallConnectorError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.dhcp import IP_ADDRESS
+from homeassistant.const import CONF_HOST, CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    CONF_SCAN_INTERVAL_CHARGING,
+    DEFAULT_SCAN_INTERVAL,
+    DEFAULT_SCAN_INTERVAL_CHARGING,
+    DOMAIN,
+    WALLCONNECTOR_DEVICE_NAME,
+    WALLCONNECTOR_SERIAL_NUMBER,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    wall_connector = WallConnector(
+        host=data[CONF_HOST], session=async_get_clientsession(hass)
+    )
+
+    version = await wall_connector.async_get_version()
+
+    return {
+        "title": WALLCONNECTOR_DEVICE_NAME,
+        WALLCONNECTOR_SERIAL_NUMBER: version.serial_number,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Tesla Wall Connector."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        super().__init__()
+        self.ip_address = None
+        self.serial_number = None
+
+    async def async_step_dhcp(self, discovery_info) -> FlowResult:
+        """Handle dhcp discovery."""
+        self.ip_address = discovery_info[IP_ADDRESS]
+        _LOGGER.info("Discovered Tesla Wall Connector at [%s]", self.ip_address)
+
+        self._async_abort_entries_match({CONF_HOST: self.ip_address})
+
+        try:
+            wall_connector = WallConnector(
+                host=self.ip_address, session=async_get_clientsession(self.hass)
+            )
+            version = await wall_connector.async_get_version()
+        except WallConnectorError as ex:
+            _LOGGER.exception(
+                "Could not read serial number from Tesla WallConnector at [%s]: [%s]",
+                self.ip_address,
+                ex,
+            )
+            return self.async_abort(reason="cannot_connect")
+
+        self.serial_number = version.serial_number
+
+        await self.async_set_unique_id(self.serial_number)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: self.ip_address})
+
+        _LOGGER.info(
+            "No entry found for wall connector with IP %s. Serial nr: %s",
+            str(self.ip_address),
+            str(self.serial_number),
+        )
+
+        self.context["description_placeholders"] = {CONF_HOST: self.ip_address}
```

Comment:
Should we show them the serial number in the discovery?

Reply:
Hmm... not sure what that would be useful for?

Reply:
I have 3 chargers (sadly v2 so I can't actually test this).

I could find the easily serial numbers but I likely wouldn't easily know which one was which if it was just by ip.

Reply:
You will have to upgrade at least one charger when this is merged! ;)

Do you have any information about how to show such information in the configuration dialog?

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/flux_led/strings.json#L3
https://github.com/home-assistant/core/blob/dev/homeassistant/components/flux_led/config_flow.py#L127

Reply:
<img width="342" alt="Screen Shot 2021-11-22 at 16 17 44" src="https://user-images.githubusercontent.com/663432/142943413-ef1cd438-5e99-4161-8c02-1382bd235bd9.png">

Reply:
Example use of `description_placeholders` as well:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/flux_led/config_flow.py#L129

Reply:
Using the examples above, I've been playing around with this on my local home assistance instance where I have the wall connector as a custom component. I haven't been able to change the title of the discovery badge from the integration title in the manifest. Do you happen to know if this is a limitation with custom_components?
I'm not able to test this on my dev instance as it does not discover anything (running in a dev container)

Reply:
Did you regenerate `en.json` with `python3 -m script.translations develop`?

Reply:
Make sure `flow_title` is in the translation file that your local language is set to for the integration.

Reply:
Thanks! That was it! I didn't know about that translations script... I assumed that keys missing from the translation files would fallback to their values in strings.json...

Reply:
What do you think about this:
![image](https://user-images.githubusercontent.com/1431687/143071090-53b11ff5-8dcc-4ce4-b10f-dc9b51c2c0d1.png)

Reply:
LGTM 👍

Reply:
Fixed in aa3092c